melior/src/ir/operation.rs

```diff
@@ -55,6 +57,53 @@ impl<'c> Operation<'c> {
         unsafe { BlockRef::from_option_raw(mlirOperationGetBlock(self.raw)) }
     }
 
+    /// Gets the number of operands.
+    pub fn operand_count(&self) -> usize {
+        unsafe { mlirOperationGetNumOperands(self.raw) as usize }
+    }
+
+    /// Gets the operand at a position.
+    pub fn operand(&self, index: usize) -> Result<Value<'c, '_>, Error> {
+        unsafe {
+            if index < self.operand_count() {
+                Ok(Value::from_raw(mlirOperationGetOperand(
+                    self.raw,
+                    index as isize,
+                )))
+            } else {
+                Err(Error::PositionOutOfBounds {
+                    name: "operation operand",
+                    value: self.to_string(),
+                    index,
+                })
+            }
+        }
+    }
+
+    /// Gets all operands.
+    pub fn operands(&self) -> Result<Vec<Value<'c, '_>>, Error> {
+        self.operands_range(0..self.operand_count())
+    }
+
+    /// Gets operands in a given range.
+    pub fn operands_range(
+        &self,
+        range: std::ops::Range<usize>,
+    ) -> Result<Vec<Value<'c, '_>>, Error> {
+        let mut operands = Vec::new();
+
+        for i in range {
+            operands.push(self.operand(i)?);
+        }
+
+        Ok(operands)
+    }
```

Comment:
I think we can delegate the "ranging" computation to users of the `operands` method (for example, they can use `.take()` and `.skip()`) as the complexity doesn't change from `O(n)` even when we do so. It's the same for the `.results()` method.

```suggestion
    /// Gets all operands.
    pub fn operands(&self) -> impl Iterator<Item = Value<'c, '_>> {
        (0..self.operand_count()).map(|index| self.operand(index).expect("valid operand index"))
    }
```

Reply:
Oh, I like this! I wasn't very happy with the `Rangle` solution myself either, this is much nicer.